### PR TITLE
PHP 8.1 compat and static analysis fixes: Fix signatures, missing returns for edge cases etc.

### PR DESCRIPTION
**Pull request by ralflang**
Nothing of this should affect functionality. This should work on PHP 7.4 as well as 8.1 and 8.2.

### CODE REVIEW

File: lib/Api.php [hunk 1]
```diff
@@ -2443,7 +2445,7 @@ public function searchTags($names, $max = 10, $from = 0,
                     if (count($result) == 0) {
                         continue;
                     } elseif (count($result) > 1) {
-                        throw new Turba_Exception(sprintf("Internal Horde Error: multiple Turba objects with same objectId %s.", $uid));
+                        throw new Turba_Exception(sprintf("Internal Horde Error: multiple Turba objects with same objectId %s.", $contact_uid));
```
Comment (ralflang) [inline lib/Api.php:2448]: Misnamed, non-existing variable.

File: lib/Api.php [hunk 2]
```diff
@@ -2550,7 +2552,7 @@ protected function _encodeContent($content)
                     $out[$attr] = $hooks->callHook(
                         'encode_attribute',
                         'turba',
-                        array($attr, $value, null, null)
+                        array($attr, $val, null, null)
```
Comment (ralflang) [inline lib/Api.php:2555]: Misnamed, non-existing variable.

File: lib/Application.php [hunk 3]
```diff
@@ -342,6 +342,7 @@ public function configSpecialValues($what)
             $addressbooks[''] = _("None");
             return $addressbooks;
         }
+        return [];
```
Comment (ralflang) [inline lib/Application.php:345]: Missing return value for edge case.

File: lib/Application.php [hunk 4]
```diff
@@ -846,6 +847,7 @@ public function download(Horde_Variables $vars)
 
             break;
         }
+        return [];
```
Comment (ralflang) [inline lib/Application.php:850]: Missing return value for edge case.

File: lib/Application.php [hunk 5]
```diff
@@ -1113,6 +1115,7 @@ public function davPutObject($collection, $object, $data)
                 $dav->addObjectMap($id, $object, $internal);
             }
         }
+        return null;
```
Comment (ralflang) [inline lib/Application.php:1118]: Missing return value for edge case.

File: lib/Application.php [hunk 6]
```diff
@@ -1140,6 +1143,7 @@ public function davDeleteObject($collection, $object)
             $dav->deleteExternalObjectId($object, $internal);
         } catch (Horde_Dav_Exception $e) {
         }
+        return true;
```
Comment (ralflang) [inline lib/Application.php:1146]: Missing return value for edge case.

File: lib/Backup/Contacts.php [hunk 7]
```diff
@@ -94,14 +94,14 @@ public function key()
 
     /**
      */
-    public function next()
+    public function next(): void
```
Comment (ralflang) [inline lib/Backup/Contacts.php:97]: Tentative return type

File: lib/Backup/Contacts.php [hunk 8]
```diff
@@ -94,14 +94,14 @@ public function key()
...
     {
         $this->_list->next();
     }
 
     /**
      */
-    public function rewind()
+    public function rewind(): void
```
Comment (ralflang) [inline lib/Backup/Contacts.php:104]: Tentative return type

File: lib/Backup/Contacts.php [hunk 9]
```diff
@@ -115,7 +115,7 @@ public function rewind()
 
     /**
      */
-    public function valid()
+    public function valid(): bool
```
Comment (ralflang) [inline lib/Backup/Contacts.php:118]: Tentative return type

File: lib/Driver/Favourites.php [hunk 10]
```diff
@@ -122,6 +122,7 @@ protected function _match($contact, $val)
             return empty($val['test']) ||
                 stristr($contact[$val['field']], $val['test']) !== false;
         }
+        return false;
```
Comment (ralflang) [inline lib/Driver/Favourites.php:125]: Missing return in edge case

File: lib/Driver/Group.php [hunk 11]
```diff
@@ -15,6 +15,8 @@
  */
 class Turba_Driver_Group extends Turba_Driver
 {
+
+    protected $_gid;
```
Comment (ralflang) [inline lib/Driver/Group.php:19]: undeclared variable used further down

File: lib/Driver/Prefs.php [hunk 12]
```diff
@@ -104,6 +104,8 @@ protected function _delete($object_key, $object_id)
      * Saves the specified object in the preferences.
      *
      * @param Turba_Object $object TODO
+     * 
+     * @return string object id
```
Comment (ralflang) [inline lib/Driver/Prefs.php:108]: Compatibility with parent driver's signature

File: lib/Driver/Prefs.php [hunk 13]
```diff
@@ -114,6 +116,7 @@ function _save($object)
         $book = $this->_getAddressBook();
         $book[$object_id] = $attributes;
         $this->_setAddressBook($book);
+        return (string) $object_id;
```
Comment (ralflang) [inline lib/Driver/Prefs.php:119]: Parent driver demands a string return

File: lib/Driver/Prefs.php [hunk 14]
```diff
@@ -139,9 +142,8 @@ protected function _getAddressBook()
      *
      * @param $addressbook TODO
      *
-     * @return TODO
      */
-    protected function _setAddressBook($addressbook)
+    protected function _setAddressBook($addressbook): void
```
Comment (ralflang) [inline lib/Driver/Prefs.php:146]: Wrong phpdoc

File: lib/Driver/Sql.php [hunk 15]
```diff
@@ -65,7 +65,7 @@ public function __construct($name = '', array $params = array())
      *
      * @return integer  The number of contacts that the user owns.
      */
-    public function count()
+    public function count(): int
```
Comment (ralflang) [inline lib/Driver/Sql.php:68]: Tentative return type

File: lib/Driver/Vbook.php [hunk 16]
```diff
@@ -35,6 +35,8 @@ class Turba_Driver_Vbook extends Turba_Driver
      */
     protected $_driver;
 
+    protected $_share;
```
Comment (ralflang) [inline lib/Driver/Vbook.php:38]: Undeclared variable

File: lib/Form/EditContactGroup.php [hunk 17]
```diff
@@ -36,9 +36,16 @@ public function __construct($vars, $contact)
         }
     }
 
-    public function renderActive($renderer, $vars, $action, $method)
+    public function renderActive(
```
Comment (ralflang) [inline lib/Form/EditContactGroup.php:39]: Parameters which are optional in parent should be optional in child.
Parameters which exist in parent should exist in child class. As they are optional, it does not hurt any existing calls.

File: lib/Form/EditContactGroup.php [hunk 18]
```diff
@@ -36,9 +36,16 @@ public function __construct($vars, $contact)
...
+        $renderer = null,
+        $vars = null,
+        $action = '',
+        $method = 'get',
+        $enctype = null,
+        $focus = true
+    )
     {
-        parent::renderActive($renderer, $vars, $action, $method);
+        parent::renderActive($renderer, $vars, $action, $method, $enctype, $focus);
```
Comment (ralflang) [inline lib/Form/EditContactGroup.php:48]: Forward the additional parameters to parent class. As they likely always have the default values, it does not break existing calls.

File: lib/List.php [hunk 19]
```diff
@@ -26,7 +26,7 @@ class Turba_List implements Countable
     /**
      * The field to compare objects by.
      *
-     * @var string
+     * @var iterable
```
Comment (ralflang) [inline lib/List.php:29]: Looking at code using this variable, it cannot be a string but an array or a list class.

File: lib/List.php [hunk 20]
```diff
@@ -254,7 +254,7 @@ protected function _cmp(Turba_Object $a, Turba_Object $b)
 
     /* Countable methods. */
 
-    public function count()
+    public function count(): int
```
Comment (ralflang) [inline lib/List.php:257]: Tentative return type

File: lib/Test.php [hunk 21]
```diff
@@ -96,7 +96,7 @@ public function appTests()
      */
     protected function _doConnectionTest($params)
     {
-        $ret .= 'server="' . htmlspecialchars($params['server']) . '" ' .
+        $ret = 'server="' . htmlspecialchars($params['server']) . '" ' .
```
Comment (ralflang) [inline lib/Test.php:99]: Cannot concat-assign to a new variable

File: lib/View/List.php [hunk 22]
```diff
@@ -511,7 +511,7 @@ public function getAddSources()
      *
      * @return integer  The number of objects in the list.
      */
-    public function count()
+    public function count(): int
```
Comment (ralflang) [inline lib/View/List.php:514]: Tentative return type

File: lib/Backup/Contacts.php (hunk 7 above)
Comment (mrubinsk) [reply lib/Backup/Contacts.php:97]: I'm a bit concerned with suddenly adding  PHP 8.1 language feature. This could cause some serious headaches for anyone running on, say, a Ubuntu 20 LTS system without having to manually go through a PHP upgrade.  An argument could definitely be made for the fact that master is a development branch, but considering the length of time since our last release and the fact that a number of people are running some sort of "master checkout", I'd like to avoid this kind of "surprise" for people, especially when we can silence the return type warnings in a BC way (for now).
